python/services/evaluationservice/dmod/evaluationservice/evaluation_service/consumers/listener.py

```diff
@@ -698,20 +629,23 @@ async def receive(self, text_data=None, **kwargs):
                     message = f"{str(self)}: '{action}' cannot be performed; " \
                               f"the following required parameters are missing: {', '.join(missing_parameters)}"
                     SOCKET_LOGGER.error(message=message)
-                    await self.send_message(result=message, event="receive", response_type="error", logger=SOCKET_LOGGER)
+                    await self.send_error(event='receive', message=message, request_id=kwargs.get(REQUEST_ID_KEY))
                     return
         except Exception as exception:
             await self.send_error(message=exception, event="receive")
             SOCKET_LOGGER.error(message=exception)
             return
 
         try:
-            await handler(action_parameters)
+            result = handler(action_parameters)
+
+            while inspect.isawaitable(result):
+                result = await result
         except Exception as exception:
             await self.send_error(message=exception, event=action)
             SOCKET_LOGGER.error(message=exception)
 
-    @required_parameters(evaluation_name="string", instructions="string")
+    @required_parameters(evaluation_name=REQUIRED_PARAMETER_TYPES.text, instructions=REQUIRED_PARAMETER_TYPES.text)
```

Comment:
I just remembered why the properties are so specific - _the values match with HTML input types_. If a function has:

```json
{
    "template_id": "number"
}
```
That will translate directly to:

```html
<label for="template_id">Template ID</label><input type="number" name="template_id" required>
```

Pydantic models may work here, but the event bus calling these functions will have to know how to deserialize everything into the right models rather than passing the input json to corresponding functions with the guarantee that the given values are there.